src/main/java/org/truffleruby/extra/RubyConcurrentMap.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2015, 2021 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.extra;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.object.Shape;
+import org.truffleruby.core.kernel.KernelNodes;
+import org.truffleruby.core.klass.RubyClass;
+import org.truffleruby.language.RubyDynamicObject;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+public class RubyConcurrentMap extends RubyDynamicObject {
+
+    public static class Key {
+
+        public final Object key;
+        private final int hashCode;
+
+        public Key(Object key, int hashCode) {
+            this.key = key;
+            this.hashCode = hashCode;
+        }
+
+        @Override
+        public int hashCode() {
+            return hashCode;
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            assert other instanceof Key;
+            final Key otherKey = (Key) other;
+
+            if (hashCode != otherKey.hashCode) {
+                return false;
+            } else {
+                // To do: It's unfortunate we're calling this behind a boundary! Can we do better?
+                final Object returnValue = KernelNodes.SameOrEqlNode.getUncached().execute(key, otherKey.key);
+                if (returnValue instanceof Boolean) {
+                    return (boolean) returnValue;
+                } else {
+                    throw new UnsupportedOperationException(returnValue.getClass().getName());
+                }
+            }
+        }
+    }
+
+    private ConcurrentHashMap<Key, Object> map;
+
+    public RubyConcurrentMap(RubyClass rubyClass, Shape shape) {
+        super(rubyClass, shape);
+    }
+
+    @TruffleBoundary
+    public void allocateMap(int initialCapacity, float loadFactor) {
+        // To do: This method is seems empty but was blacklisted on CI
```

Comment:
@eregon why is this empty method blacklisted?

Reply:
`new ConcurrentHashMap<>` is not PE-safe (the constructor is not trivial), so we do need the boundary (and we can remove the comment)

Reply:
It's just sets some fields and does some arithmetic, doesn't it?

```java
    public ConcurrentHashMap(int initialCapacity,
                             float loadFactor, int concurrencyLevel) {
        if (!(loadFactor > 0.0f) || initialCapacity < 0 || concurrencyLevel <= 0)
            throw new IllegalArgumentException();
        if (initialCapacity < concurrencyLevel)   // Use at least as many bins
            initialCapacity = concurrencyLevel;   // as estimated threads
        long size = (long)(1.0 + (long)initialCapacity / loadFactor);
        int cap = (size >= (long)MAXIMUM_CAPACITY) ?
            MAXIMUM_CAPACITY : tableSizeFor((int)size);
        this.sizeCtl = cap;
    }
```

Reply:
That exception is reachable, and fillInStackTrace is blacklisted. We don't want to PE any of that.

Reply:
It seems to PE just fine?

All the sizing logic constant folds if you're using the defaults `loop { TruffleRuby::ConcurrentMap.new }`, which almost everyone is.

If you aren't using the defaults `loop { TruffleRuby::ConcurrentMap.new(initial_capacity: rand(10), load_factor: rand) }` it still PEs to reasonable code.

Reply:
`Throwable#fillInStackTrace` is blacklisted for PE here:
https://github.com/oracle/graal/blob/d0371a3e07ade18d800ad9ec22e0900320a9be19/substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleFeature.java#L733
And all ConcurrentHashMap methods are blacklisted for PE here:
https://github.com/oracle/graal/blob/d0371a3e07ade18d800ad9ec22e0900320a9be19/substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleFeature.java#L776
That check is very useful in general, so we want to keep it.

This is arbitrary JDK code, which is even likely to change between JDK versions, so it's not something we can PE safely.